File: snake-ai-system-frontend/src/api/snake/speciesImages.ts
import { snakeClient } from '../snake'

export interface SpeciesImage {
  id: number
  species_id: number
  image_url: string
  thumbnail_url?: string
  caption?: string
  photographer?: string
  image_type: string
  sort_order: number
  is_cover: boolean
  created_at: string
}

export interface SpeciesImageCreate {
  species_id: number
  image_url: string
  thumbnail_url?: string
  caption?: string
  photographer?: string
  image_type?: string
  sort_order?: number
  is_cover?: boolean
}

export interface SpeciesImageUpdate {
  image_url?: string
  thumbnail_url?: string
  caption?: string
  photographer?: string
  image_type?: string
  sort_order?: number
  is_cover?: boolean
}

export const speciesImagesApi = {
  list: (speciesId: number) =>
    snakeClient.get<SpeciesImage[]>(`/species/${speciesId}/images`).then((res) => res.data),

  create: (speciesId: number, data: Omit<SpeciesImageCreate, 'species_id'>) =>
    snakeClient.post<SpeciesImage>(`/species/${speciesId}/images`, data).then((res) => res.data),

  update: (imageId: number, data: SpeciesImageUpdate) =>
    snakeClient.patch<SpeciesImage>(`/images/${imageId}`, data).then((res) => res.data),

  delete: (imageId: number) => snakeClient.delete(`/images/${imageId}`),
}

File: backend/app/routers/species_images.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.services.database import get_db
from app.services.models import Species, SpeciesImages
from app.services.db_utils import commit_or_409
from app.services.security import get_current_reviewer_user
import app.services.schemas as schemas

router = APIRouter(tags=["物种图片"])


@router.post(
    "/species/{species_id}/images",
    response_model=schemas.SpeciesImageResponse,
    status_code=201,
)
async def create_species_image(
    species_id: int,
    data: schemas.SpeciesImageCreate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_reviewer_user),
):
    species = await db.get(Species, species_id)
    if not species:
        raise HTTPException(404, "物种不存在")
    image = SpeciesImages(**{**data.model_dump(exclude={"species_id"}), "species_id": species_id})
    db.add(image)
    await commit_or_409(db)
    await db.refresh(image)
    return image


@router.get(
    "/species/{species_id}/images",
    response_model=list[schemas.SpeciesImageResponse],
)
async def list_species_images(species_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(SpeciesImages)
        .where(SpeciesImages.species_id == species_id)
        .order_by(SpeciesImages.sort_order)
    )
    return result.scalars().all()


@router.patch("/images/{image_id}", response_model=schemas.SpeciesImageResponse)
async def update_image(
    image_id: int,
    data: schemas.SpeciesImageUpdate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_reviewer_user),
):
    image = await db.get(SpeciesImages, image_id)
    if not image:
        raise HTTPException(404, "图片不存在")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(image, key, value)
    await commit_or_409(db)
    await db.refresh(image)
    return image


@router.delete("/images/{image_id}", status_code=204)
async def delete_image(
    image_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_reviewer_user),
):
    image = await db.get(SpeciesImages, image_id)
    if not image:
        raise HTTPException(404, "图片不存在")
    await db.delete(image)
    await db.commit()
